### PR TITLE
Dataframe improvements

### PR DESCRIPTION
**Pull request by sg-s**
## changes 

- fixes a bug where empty rows were skipped
- types many columns in pandas dataframes

### CODE REVIEW

File: src/managed_data/api.py [hunk 1]
```diff
@@ -298,12 +298,14 @@ def get_dataframe(
         data[column["id"]] = []
 
     for row in rows:
-        if "fields" not in row.keys():
-            continue
-
         data[row_id].append(row["hid"])
```
Comment (sg-s) [inline src/managed_data/api.py:301]: this caused empty rows to be skipped

File: src/managed_data/api.py [hunk 2]
```diff
@@ -298,12 +298,14 @@ def get_dataframe(
...
         data["Validation Status"].append(row["validationStatus"])
 
+        if "fields" not in row.keys():
+            for column in columns:
+                data[column["id"]].append(None)
+            continue
+
```
Comment (sg-s) [inline src/managed_data/api.py:308]: if we have an empty row, we append None to all columns and hope for the best

File: src/managed_data/api.py [hunk 3]
```diff
@@ -315,10 +317,13 @@ def get_dataframe(
...
-            if column["cardinality"] == "many":
-                data[column["id"]].append(value)
+            if value is None:
+                data[column["id"]].append(None)
             else:
-                data[column["id"]].extend(value)
+                if column["cardinality"] == "many":
+                    data[column["id"]].append(value)
+                else:
+                    data[column["id"]].extend(value)
```
Comment (sg-s) [inline src/managed_data/api.py:326]: values can now be None, so we need to handle that case

File: src/managed_data/api.py [hunk 4]
```diff
@@ -367,10 +372,10 @@ def _parse_column_value(
     field = [field for field in fields if field["columnId"] == column["id"]]
 
     if len(field) == 0:
-        return [None]
+        return None
 
     if "value" not in field[0].keys():
-        return [None]
+        return None
```
Comment (sg-s) [inline src/managed_data/api.py:378]: i don't remember why i did this, it seems simpler to return None

File: src/managed_data/api.py [hunk 5]
```diff
@@ -421,17 +426,36 @@ def _type_and_cleanup_dataframe(
...
+            df[col_id] = pd.to_datetime(df[col_id])
+
+        # special treatment for string columns
+        if column["type"] in ["file", "text"]:
+            df[col_id] = df[col_id].astype("string")
+
+        if column["type"] == "boolean":
+            df[col_id] = df[col_id].astype("boolean")
+
+        # special treatment of Select columns
```
Comment (sg-s) [inline src/managed_data/api.py:445]: better typing. boolean is nullable.

File: src/managed_data/api.py [hunk 6]
```diff
@@ -421,17 +426,36 @@ def _type_and_cleanup_dataframe(
...
 
     # rename columns
     df = df.rename(columns=column_mapper)
 
+    # special treatment for validation status
+    categories = set(df["Validation Status"])
+    df["Validation Status"] = pd.Categorical(
+        df["Validation Status"], categories=categories
+    )
+
```
Comment (sg-s) [inline src/managed_data/api.py:458]: this should work for any validation status, without needing to know what they are

**Review by sg-s — COMMENTED** [review]
self-review

File: src/managed_data/api.py (hunk 2 above)
Comment (kennyjwilli) [reply src/managed_data/api.py:308]: Why? Can a dataframe row not just be empty?

File: src/managed_data/api.py [hunk 7]
```diff
@@ -421,17 +426,36 @@ def _type_and_cleanup_dataframe(
     for column in columns:
         column_mapper[column["id"]] = column["name"]
 
-    # special treatment of Select columns
     for column in columns:
         col_id = column["id"]
 
+        print(col_id)
+        print(column["type"])
```
Comment (kennyjwilli) [inline src/managed_data/api.py:433]: Remove

**Review by kennyjwilli — APPROVED** [review]
I'd say just remove the prints and then it's good to go 👍🏻

File: src/managed_data/api.py (hunk 2 above)
Comment (sg-s) [reply src/managed_data/api.py:308]: no. we have to put something in it

File: src/managed_data/api.py (hunk 7 above)
Comment (sg-s) [reply src/managed_data/api.py:433]: goot catch, thanks!

**Pull request merged**